### ISSUE TITLE
Regression: Switching keyboard layouts results in corrupted keyboard output

### ISSUE THREAD

**Issue by juj** [microsoft/terminal#17187]
### Windows Terminal version

1.19.10821.0

### Windows build number

10.0.22631.3447

### Other Software

N/A

### Steps to reproduce

1. Open command prompt
2. Change to another keyboard layout in taskbar (e.g. from English to Finnish)
3. Bug 1: open the keyboard layout UI in task bar, and observe that the keyboard layout is showing conflicting state
4. Type in chars to the command prompt window
5. Bug 2: observe that typed keys are incorrect (they do not follow the Finnish keymap)
6. Close the command prompt window
7. Change the keyboard layout to Finnish
8. Open a new command prompt, so that this prompt starts up immediately in the Finnish keyboard layout
9. Type in chars to the command prompt window
10. Observe that this time they keys are indeed correct for Finnish keyboard layout, unlike in step 4.

So it looks like starting up command prompt in the desired keyboard layout produces correct keys to that layout, but starting prompt in another layout and then attempting to change the layout while command prompt is running, results in corrupted layout.

Showcase video of the bug in action: https://youtu.be/LDEJsrhHgRo

### Actual Behavior

Incorrect keys are output.

**Comment by zadjii-msft** [comment]
You know what, we just published a major update to our IME implementation in the nightly Canary branch. It was rewritten from the ground up. Can you try that out, and see if that repros this behavior/? You can get it from: https://aka.ms/terminal-canary-installer
If you already have the Canary build installed, you can use this link to force an update.

**Comment by juj** [comment]
Thanks for the quick reply. Before I try it out, is the canary installation somehow separate (and uninstallable?) so that it won't overwrite my regular command prompt? Or is there a way to roll back to the previous command prompt from the canary one?

**Comment by lhecker** [comment]
Yes it's a completely separate, packaged application. It'll not conflict with your other Windows Terminal installations. And if you uninstall it, it'll be gone as if it was never installed. :)

I don't think Finnish relies on IME/TSF, but I think it's worth a try regardless, since there's a number of other input related changes in there.

**Comment by juj** [comment]
Ok, trying out the Canary version, the layout issue does not reproduce there.

I've also tried a couple of other applications, and the layout issue does not occur in any of those either: Notepad, Sublime Text, Firefox and Chrome web browser, Paint.NET, they are all good. Only cmd.exe is giving this keyboard layout issue.

![image](https://github.com/microsoft/terminal/assets/225351/6b7cd760-39ad-4180-bcf6-eda57cb6b12a)

**Comment by juj** [comment]
Not much info to glean about the Text Input Management Service...

![image](https://github.com/microsoft/terminal/assets/225351/ad70b2a9-28d8-4951-848f-dd29b525f156)

**Comment by carlos-zamora** [comment]
Thanks for confirming that it doesn't repro on Canary. This should be available now in WT Preview v1.21. Closing as fixed.

**Comment by DHowett** [comment]
You should be able to install Terminal 1.21 via winget or the Releases page in this repository. We are rolling it out slowly on account of a platform issue in Windows 11 23H2 (10.0.226xx.0), as we've measured an increase in hanging sessions.

**Comment by juj** [comment]
It turns out this issue is still happening, but the characters that are getting printed have changed. I think this bug should be reopened.

Microsoft Windows [Version 10.0.26100.2605]

Windows Terminal
Version: 1.21.10351.0

Or should I open a new issue?

**Comment by lhecker** [comment]
I think since it's just us here, we can just reopen this issue. This lets us keep its entire history. In what way does it behave now?

**Comment by juj** [comment]
Here is a video: https://youtu.be/BFI-9UfRjNk

The behavior is quit the same as before. The only difference I notice is that the incorrect characters that are displayed are different than in the first video.

**Comment by juj** [comment]
Friendly ping: I wonder if this issue could be looked at?

I got a new PC, so this issue is reproducing on two different Windows 11 systems I have now. It is causing a lot of annoyance, and a major regression from Windows 10.

Setup is a rather simple dual keyboard layout that I've used as long as I remember, since Windows NT days or similar.

![Image](https://github.com/user-attachments/assets/bee674ef-eca5-4b8a-ab78-07c8ab5b6b04)

Windows 11 Pro 23H2
Microsoft Windows [Version 10.0.22631.5335]
Windows Terminal
Version: 1.22.11141.0

**Comment by juj** [comment]
Ping, any updates here? This issue is still happening.

Microsoft Windows [Version 10.0.26100.4351]
Windows Terminal Version: 1.22.12111.0
